### PR TITLE
Remove dependency on TargetIsa from Wasmtime crates 

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit started off by deleting the `cranelift_codegen::settings`
reexport in the `wasmtime-environ` crate and then basically played
whack-a-mole until everything compiled again. The main result of this is
that the `wasmtime-*` family of crates have generally less of a
dependency on the `TargetIsa` trait and type from Cranelift. While the
dependency isn't entirely severed yet this is at least a significant
start.

This commit is intended to be largely refactorings, no functional
changes are intended here. The refactorings are:

* A `CompilerBuilder` trait has been added to `wasmtime_environ` which
  server as an abstraction used to create compilers and configure them
  in a uniform fashion. The `wasmtime::Config` type now uses this
  instead of cranelift-specific settings. The `wasmtime-jit` crate
  exports the ability to create a compiler builder from a
  `CompilationStrategy`, which only works for Cranelift right now. In a
  cranelift-less build of Wasmtime this is expected to return a trait
  object that fails all requests to compile.

* The `Compiler` trait in the `wasmtime_environ` crate has been souped
  up with a number of methods that Wasmtime and other crates needed.

* The `wasmtime-debug` crate is now moved entirely behind the
  `wasmtime-cranelift` crate.

* The `wasmtime-cranelift` crate is now only depended on by the
  `wasmtime-jit` crate.

* Wasm types in `cranelift-wasm` no longer contain their IR type,
  instead they only contain the `WasmType`. This is required to get
  everything to align correctly but will also be required in a future
  refactoring where the types used by `cranelift-wasm` will be extracted
  to a separate crate.

* I moved around a fair bit of code in `wasmtime-cranelift`.

* Some gdb-specific jit-specific code has moved from `wasmtime-debug` to
  `wasmtime-jit`.

~~Note that this is built on https://github.com/bytecodealliance/wasmtime/pull/3176, but I wanted to keep all these refactorings ideally bite-ish sized to be more easily digestable~~

### CODE REVIEW

File: cranelift/wasm/src/environ/spec.rs [hunk 1]
```diff
@@ -702,7 +702,7 @@ pub trait FuncEnvironment: TargetEnvironment {
 /// An object satisfying the `ModuleEnvironment` trait can be passed as argument to the
 /// [`translate_module`](fn.translate_module.html) function. These methods should not be called
 /// by the user, they are only for `cranelift-wasm` internal use.
-pub trait ModuleEnvironment<'data>: TargetEnvironment {
+pub trait ModuleEnvironment<'data> /*: TargetEnvironment*/ {
```
Comment (bjorn3) [inline cranelift/wasm/src/environ/spec.rs:705]: Just a reminder to either remove or uncomment this code.

File: crates/environ/src/module_environ.rs [hunk 2]
```diff
@@ -378,15 +366,15 @@ impl<'data> ModuleEnvironment<'data> {
...
+//     }
 
-    fn reference_type(&self, ty: cranelift_wasm::WasmType) -> ir::Type {
-        crate::reference_type(ty, self.pointer_type())
-    }
-}
+//     fn reference_type(&self, ty: cranelift_wasm::WasmType) -> ir::Type {
+//         crate::reference_type(ty, self.pointer_type())
+//     }
+// }
```
Comment (bjorn3) [inline crates/environ/src/module_environ.rs:377]: Same

File: crates/jit/src/compiler.rs [hunk 3]
```diff
@@ -41,31 +40,30 @@ pub enum CompilationStrategy {
...
+    pub fn builder(strategy: CompilationStrategy) -> Box<dyn CompilerBuilder> {
+        match strategy {
+            CompilationStrategy::Auto | CompilationStrategy::Cranelift => {
+                wasmtime_cranelift::builder()
+            }
+            #[cfg(feature = "lightbeam")]
+            CompilationStrategy::Lightbeam => unimplemented!(),
+        }
+    }
+    /// TODO
```
Comment (bjorn3) [inline crates/jit/src/compiler.rs:59]:
```suggestion
    /// Construct a new `Compiler`.
```

File: crates/wasmtime/src/config.rs [hunk 4]
```diff
@@ -1210,23 +1175,22 @@ impl Config {
...
-    pub(crate) fn target_isa_with_reference_types(&self) -> Box<dyn TargetIsa> {
-        let mut flags = self.flags.clone();
-        flags.set("enable_safepoints", "true").unwrap();
-        self.isa_flags.clone().finish(settings::Flags::new(flags))
-    }
+    // pub(crate) fn target_isa_with_reference_types(&self) -> Box<dyn TargetIsa> {
+    //     let mut flags = self.flags.clone();
+    //     flags.set("enable_safepoints", "true").unwrap();
+    //     self.isa_flags.clone().finish(settings::Flags::new(flags))
+    // }
```
Comment (bjorn3) [inline crates/wasmtime/src/config.rs:1188]: Same

File: crates/runtime/src/instance/allocator/pooling.rs [hunk 5]
```diff
@@ -1088,7 +1087,6 @@ mod test {
             style: TableStyle::CallerChecksSignature,
             table: Table {
                 wasm_ty: WasmType::FuncRef,
-                ty: TableElementType::Func,
```
Comment (pchickey) [inline crates/runtime/src/instance/allocator/pooling.rs:1091]: I am grateful that we have gotten rid of all these redundant duplications!

**Review by pchickey — APPROVED** [review]

**Pull request merged**